pkg/scheduler/objects/sorters.go

```diff
@@ -65,8 +65,10 @@ func sortQueuesByPriorityAndFairness(queues []*Queue) {
 		if lPriority < rPriority {
 			return false
 		}
-		comp := resources.CompUsageRatioSeparately(l.GetAllocatedResource(), l.GetGuaranteedResource(),
-			r.GetAllocatedResource(), r.GetGuaranteedResource())
+
```

Comment:
I think this function needs some work as long as we're rewriting it. We are doing the expensive recursive computation of resources on every comparison. Since we already have a slice of queues, it would be better to precompute a slice of resources computed on each one, and then use that in the comparison function so that we only create the resource maps once per sort run instead of once per comparison.

Reply:
Wilfred also pointed out in the community meeting that we may also be able to optimize the parent hierarchy separately as all the compared queues will have the same parent. If we do that, we do need to be sure that the parent is not the root.

Reply:
Currently on vacation, but I will add this optimization when I return.

Reply:
pushed.  This should compute GetFairMaxResource O(n) times instead of NLog(N).

Reply:
This is better, but to Wilfred's point from the community meeting, check out the callers of `queue.sortQueues()` -- the are in queue.go under `GetQueueOutstandingRequests()`, `TryAllocate()`, `TryPlaceholderAllocate()` and `TryReservedAllocate()`. Each computes `queue.getHeadRoom()` at the start which is used for further processing. Since these methods are all recursive from the root queue down, you can compute the parent values at each step and then pass the results into the `sortQueues()` function. This avoids having to recompute all the parent queue values during sorting.

Reply:
I'm okay with deferring that refactor until a follow-up PR. @wilfred-s what do you think?

Reply:
We can defer that to the next release. It is not just a question of performance but also a question about correctness. Since the queue values can change it might be the case that a parents' setting changes. If that happens between the child checks, specially when you have a larger number of children, you might not have the exact same inputs for each child computation.
The chance is low but we should try an prevent that.
The change would not be in the sorter but in the queue code when we build the `fairMaxResources` slice